### PR TITLE
Add user access function, edit bot message, confirm function, and message manipulation

### LINKED ISSUE #16: A member with a specific role can send or edit a message through the bot
The user who has a certain role can make the bot send a message, and cannot see the option if they don't have the role.

### PR DESCRIPTION
**Pull request by ZuliusRoolf**
This pull request includes the following changes:

- Added a user access function in the utility module to check if a user has a specific role or higher.

- Added the ability to edit a bot message.

- Added a confirm function to check if a user has confirmed an action by typing 'yes' or 'no'.

- Added message manipulation functions.

These changes enhance the functionality of the software by providing access control, message editing, and confirmation capabilities.

### CODE REVIEW

**Review by ZuliusRoolf — COMMENTED** [review]
Perhaps a bit over engineered, tell me if I should break the command arguments into smaller functions instead. Maybe even create a class would be good.

For now it works, it has been manually tested so that all features works. Checked somewhat if errors was caught correctly when inserting wrong values.

**Comment by andreas-bauer** [comment]
I refacatored the `utility.py` file. Have a look at the changes.

File: discord-bot/main.py [hunk 1]
```diff
@@ -52,8 +58,91 @@ async def github(interaction: discord.Interaction):
...
+async def confirm(interaction: discord.Interaction):
+    await interaction.response.send_message("Amazing", ephemeral=True)
+    if await user_has_confirmed(interaction, bot):
+        await interaction.followup.send("Confirmed", ephemeral=True)
+
+@bot.tree.command(description="Command me to send, change or delete my messages.")
+@app_commands.describe(action="Select the type of action you want to perform.")
+@app_commands.describe(target="Select the target ID to send, edit or delete.")
+@app_commands.describe(source="Copy or move from source message ID to target channel ID.")
+async def message(
```
Comment (andreas-bauer) [inline discord-bot/main.py:71]: This function is too nested and needs to be refactored.

The specific problem is called Nested Conditional and should be replaced with Guard Clauses.

Bad:
```js
someRequest(req, res) {
   const config = someOperation();
   if (!config) {
       res.status(404).send('Config not found')
     } else {
       res.status(200).json(config)
   }
}
```

Better:
```js
someRequest(req, res) {
const config = someOperation();
    if (!config) {
        // exception handling path
        res.status(404).send('Config not found')
        return
    }

    // happy path (the expected result/operation if everything works)
    res.status(200).json(config)
    }
}
```

For this function:
```python
if await user_has_access(interaction, ADMIN_ROLE_ID, minimum=True):
   // do something
```

```python
has_access =  user_has_access(interaction, ADMIN_ROLE_ID, minimum=True)
if not has_access:
   // handle problem

// continue with happy path

```

File: discord-bot/main.py [hunk 2]
```diff
@@ -52,8 +58,91 @@ async def github(interaction: discord.Interaction):
...
+    interaction: discord.Interaction,
+    action: Literal["send", "edit", "delete", "copy", "move"],
+    target: Optional[str] = None,
+    source: Optional[str] = None):
+    if await user_has_access(interaction, ADMIN_ROLE_ID, minimum=True):
+        def check(m):
```
Comment (andreas-bauer) [inline discord-bot/main.py]: `check()` should be defined outside of the function or as lambda.
Further, the function name `check` is ambiguous and should be replaced with a better name, e.g., `check_user_permission()`.

File: discord-bot/main.py [hunk 3]
```diff
@@ -52,8 +58,91 @@ async def github(interaction: discord.Interaction):
...
+            except ValueError:
+                await interaction.followup.send("Invalid channel ID.", ephemeral=True)
+                return
+            await interaction.followup.send("Please enter the content for the message:", ephemeral=True)
+            new_message = await bot.wait_for('message', check=check)
+            await target_channel.send(new_message.content)
+            await interaction.followup.send(f"Message sent to {target_channel.mention}", ephemeral=True)
+            await new_message.delete()
+
+        elif action in ['edit', 'delete']:
```
Comment (andreas-bauer) [inline discord-bot/main.py]: The handling of both `edit` and `delete` is very nested and hard to understand.

Split up the two actions, even if it introduces some code duplication. 

```python
if action == "send":
   // handle send

if action == "edit":
   // handle edit

if action == "delete":
   // handle delete
```

File: discord-bot/main.py [hunk 4]
```diff
@@ -52,8 +58,91 @@ async def github(interaction: discord.Interaction):
...
+                new_message = await bot.wait_for('message', check=check)
+                await target_message.edit(content=new_message.content)
+                await interaction.followup.send(f"Message edited: {target_message.jump_url}", ephemeral=True)
+                await new_message.delete()
+            elif action == 'delete':
+                if await user_has_confirmed(interaction, bot, content=f"Are you sure you want to delete {target_message.jump_url}?"):
+                    await interaction.followup.send(f"Message deleted: \n{target_message.content}", ephemeral=True)
+                    await target_message.delete()
+
+        elif action in ['copy', 'move']:
```
Comment (andreas-bauer) [inline discord-bot/main.py]: Similar to `edit` and `delete`. Split up the functionality.

**Review by andreas-bauer — CHANGES_REQUESTED** [review]

**Review by ZuliusRoolf — COMMENTED** [review]
Decided to move everything to its own class. If there are any heavy commands in the future they can be defined there. Hopefully this is more readable :)

**Review by andreas-bauer — APPROVED** [review]
Looks much better now

**Pull request merged**